[Minor enhancement]  Startup txindex after ActivateBestChain

ptschip opened the pull request:
In some cases, when someone is doing IBD on spinning disk with txindex enabled, then stops halfway through IBD and restarts later, then the ActivateBestChain step can take a very long time because of the disk contention with txindex needing to do a lot of compaction from the previous IBD run.  So here we startup txindex only after ActiveBestChain is complete which allows the node to startup much faster.

ptschip committed:
Startup txindex near the end of the node initialization

In a case where an IBD is halted then restarted we can end up having
a great deal of txindex compaction to do which can slow down node
startup significantly.  If we start txindex near the end, we will
make the node available much faster which txindex runs in the
background.

ptschip committed:
Check for valid g_txindex pointer rather than fTxIndex flag

We're going to use the pointer so we should check for the pointer
validitity rather than relying on whether the fTxIndex flag is set
correctly. Also we can check for pointer validity by using
IsTxIndexReady().

sickpig commented:
@ptschip txindex.py is failing 

```
txindex.py: Pass: False, Duration: 49 s
##################################################
- Process /home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/../qa/rpc-tests/txindex.py --srcdir /home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/src --coveragedir /tmp/coveragejy0x4jow --no-ipv6-rpc-listen --portseed=618 return code: 0
- stdout --------------------------------------------------
 2020-02-16 22:40:35,666.INFO: Random seed: 1581892835
Initializing test directory /tmp/test_txindexpy_49b32hej
2020-02-16 22:40:36,953.INFO: Started /home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/src/bitcoind 0 as pid 24704 at 127.0.0.1:15944 dir /tmp/test_txindexpy_49b32hej/node0  
2020-02-16 22:40:37,728.INFO: Started /home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/src/bitcoind 1 as pid 24809 at 127.0.0.1:15945 dir /tmp/test_txindexpy_49b32hej/node1  
2020-02-16 22:40:37,843.INFO: Mining blocks...
2020-02-16 22:40:40,626.INFO: Checking non txindex on node0...
2020-02-16 22:40:42,041.INFO: Checking txindex on node1...
2020-02-16 22:40:43,488.INFO: Checking invalid tx...
2020-02-16 22:40:51,803.INFO: Checking tx added to mempool...
2020-02-16 22:40:52,858.INFO: Checking mined tx...
2020-02-16 22:40:54,953.INFO: Mining blocks...
2020-02-16 22:40:56,090.INFO: Checking txindex on node1...
2020-02-16 22:40:57,598.INFO: Restarting...
2020-02-16 22:41:00,045.INFO: Started /home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/src/bitcoind 0 as pid 26240 at 127.0.0.1:15944 dir /tmp/test_txindexpy_49b32hej/node0  
2020-02-16 22:41:00,300.INFO: Started /home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/src/bitcoind 1 as pid 26264 at 127.0.0.1:15945 dir /tmp/test_txindexpy_49b32hej/node1  
2020-02-16 22:41:00,424.INFO: Mining more blocks...
2020-02-16 22:41:01,850.INFO: Restarting...
2020-02-16 22:41:04,286.INFO: Started /home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/src/bitcoind 0 as pid 26396 at 127.0.0.1:15944 dir /tmp/test_txindexpy_49b32hej/node0  
2020-02-16 22:41:04,542.INFO: Started /home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/src/bitcoind 1 as pid 26450 at 127.0.0.1:15945 dir /tmp/test_txindexpy_49b32hej/node1  
2020-02-16 22:41:04,668.INFO: Checking non txindex on node0...
2020-02-16 22:41:13,356.INFO: Checking txindex on node1...
2020-02-16 22:41:14,254.INFO: Restarting...
2020-02-16 22:41:16,590.INFO: Started /home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/src/bitcoind 0 as pid 27013 at 127.0.0.1:15944 dir /tmp/test_txindexpy_49b32hej/node0  
2020-02-16 22:41:17,939.INFO: Started /home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/src/bitcoind 1 as pid 27040 at 127.0.0.1:15945 dir /tmp/test_txindexpy_49b32hej/node1  
2020-02-16 22:41:18,064.INFO: Checking txindex on node0...
2020-02-16 22:41:23,641.ERROR: Assertion failed: getrawtransaction failed
2020-02-16 22:41:23,641.INFO: Stopping nodes
2020-02-16 22:41:24,611.INFO: Not cleaning up dir /tmp/test_txindexpy_49b32hej
2020-02-16 22:41:24,612.ERROR: Failed
```

ptschip committed:
Do not start the txindexing process until the node has fully synced

ptschip committed:
Remove log spam when checking synced status

gandrewstone commented on src/init.cpp:
```diff
@@ -1573,6 +1562,18 @@ bool AppInit2(Config &config, thread_group &threadGroup)
+    if (fTxIndex)
+    {
+        uiInterface.InitMessage(_("Starting txindex"));
+        auto txindex_db = new TxIndexDB(cacheConfig.nTxIndexCache, false, GetBoolArg("-reindex", DEFAULT_REINDEX));
```
why not use fReindex as the 3rd arg?

ptschip replied:
I believe this was correct but I added some clarifying code and comments.

> why not use fReindex as the 3rd arg?

I thinks this is correct but I added some clarifying comments and code.

ptschip committed:
Clarify the code and add comments when we startup the txindex database

We don't want to rely of the fReindex flag to determine whether to wipe
and rebuild the txindex because, in the case of a very small reindex, it
could already be complete by the time we get to the end of the node
startup where we initialize the txindex, and therefore the txindex would
not get rebuilt. So instead we determine whether the config parameters
had asked for a -reindex and if so then also rebuild the txindex.
